services/placement/filter/mirrored.go

```diff
@@ -0,0 +1,435 @@
+}
+
+func (f *mirroredFilter) FilterInitialInstances(
+	candidates []services.PlacementInstance,
+	rf int,
+) ([]services.PlacementInstance, error) {
+	candidates, err := getValidCandidates(placement.NewPlacement(), candidates, f.opts)
+	if err != nil {
+		return nil, err
+	}
+
+	weightToHostMap, err := hostsGroupedByWeight(candidates)
+	if err != nil {
+		return nil, err
+	}
+
+	var pairs = make([][]services.PlacementInstance, 0, len(candidates))
+	for _, hosts := range weightToHostMap {
+		pairedHosts := pairHosts(hosts, rf)
+		if len(pairedHosts) == 0 {
+			continue
+		}
+
+		pairedInstances, err := pairInstances(pairedHosts)
+		if err != nil {
+			return nil, err
+		}
+
+		pairs = append(pairs, pairedInstances...)
+	}
+
+	if len(pairs) == 0 {
+		return nil, errNoValidMirrorInstance
+	}
+
+	return assignShardsToPairedInstances(pairs, placement.NewPlacement()), nil
+}
+
+func (f *mirroredFilter) FilterAddingInstances(
+	candidates []services.PlacementInstance,
+	p services.Placement,
+) ([]services.PlacementInstance, error) {
+	candidates, err := getValidCandidates(p, candidates, f.opts)
+	if err != nil {
+		return nil, err
+	}
+
+	weightToHostMap, err := hostsGroupedByWeight(candidates)
+	if err != nil {
+		return nil, err
+	}
+
+	var pairs = make([][]services.PlacementInstance, 0, len(candidates))
+	for _, hosts := range weightToHostMap {
+		pairedHosts := pairHosts(hosts, p.ReplicaFactor())
+		if len(pairedHosts) == 0 {
+			continue
+		}
+
+		pairs, err = pairInstances(pairedHosts[:1])
+		if err != nil {
+			return nil, err
+		}
+		break
+	}
+
+	if len(pairs) == 0 {
+		return nil, errNoValidMirrorInstance
+	}
+
+	return assignShardsToPairedInstances(pairs, p), nil
+}
+
+// FindReplaceInstances for mirror supports replacing multiple instances from one host.
+// Two main use cases:
+// 1, find a new host from a pool of hosts to replace a host in the placement
+// 2, back out of a replacement, both leaving and adding host are still in the placement
+func (f *mirroredFilter) FilterReplaceInstances(
+	candidates []services.PlacementInstance,
+	leavingInstanceIDs []string,
+	p services.Placement,
+) ([]services.PlacementInstance, error) {
+	candidates, err := getValidCandidates(p, candidates, f.opts)
+	if err != nil {
+		return nil, err
+	}
+
+	leavingInstances := make([]services.PlacementInstance, 0, len(leavingInstanceIDs))
+	for _, id := range leavingInstanceIDs {
+		leavingInstance, exist := p.Instance(id)
+		if !exist {
+			return nil, errInstanceAbsent
+		}
+		leavingInstances = append(leavingInstances, leavingInstance)
+	}
+
+	// Validate leaving instances.
+	var (
+		h     host
+		ssIDs = make(map[string]struct{}, len(leavingInstances))
+	)
+	for _, instance := range leavingInstances {
+		if h.name == "" {
+			h.name = instance.Hostname()
+			h.weight = instance.Weight()
+			h.rack = instance.Rack()
+			h.portToInstance = make(map[string]services.PlacementInstance, len(leavingInstances))
+		}
+		if h.weight != instance.Weight() {
```

Comment:
You can still return an error from the `Equals` method and include details in the error, no?